hawkmoth/parser.py

```diff
@@ -413,7 +421,23 @@ def parse(filename, domain='c', clang_args=None):
         result.add_child(ds)
 
     for cursor in tu.cursor.get_children():
-        if cursor.hash in comments:
+
+        # Identify `extern "C"` and `extern "C++"` blocks and recursively parse
+        # their contents. Only `extern "C"` is of any relevance in choosing a
+        # different domain.
+        # For some reason, the Python bindings don't return the cursor kind
+        # LINKAGE_SPEC as one would expect, so we need to do it the hard way.
+        if cursor.kind == CursorKind.UNEXPOSED_DECL:
+            tokens = cursor.get_tokens()
+            if next(tokens).spelling == 'extern':
+                nested_domain = 'c' if next(tokens).spelling == '"C"' else domain
+                for c in cursor.get_children():
+                    if c.hash in comments:
+                        result.add_children(_recursive_parse(nested_domain,
+                                                             comments, errors, c, 0))
```

Comment:
Does this not document the children of `UNEXPOSED_DECL` twice? For the first time as the children of `UNEXPOSED_DECL`, and the second time as the children of `tu.cursor.get_children()`?

Reply:
Also, would be nice to have a helper for identifying the `extern "<lang>"` part instead of inlining it at the highest level parser function. Can be fixed later too, though.

Reply:
> Does this not document the children of `UNEXPOSED_DECL` twice? For the first time as the children of `UNEXPOSED_DECL`, and the second time as the children of `tu.cursor.get_children()`?

No. The 2nd loop only recurses onto cursors that have a docstring associated with them and some cursors never get docstrings of their own according to our skip rules. `UNEXPOSED_DECL` is one of them and any preceding docstring is treated as a top level comment.

In fact I'm demoing this in some actual code now and one issue I have is kind of related: nothing shows up in the output because I'm not looking into namespaces, which will probably require a similar handling here.